packages/opentelemetry-core/src/platform/BaseAbstractPlugin.ts

```diff
@@ -0,0 +1,59 @@
+/*!
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Tracer,
+  Plugin,
+  Logger,
+  PluginConfig,
+  TracerProvider,
+  PluginInternalFiles,
+} from '@opentelemetry/api';
+
+/** This class represent the base to patch plugin. */
+export abstract class BaseAbstractPlugin<T> implements Plugin<T> {
+  abstract readonly moduleName: string; // required for internalFilesExports
+  supportedVersions?: string[];
+  readonly version?: string; // required for internalFilesExports
+  protected readonly _basedir?: string; // required for internalFilesExports
+
+  protected _moduleExports!: T;
+  protected _internalFilesExports!: { [module: string]: unknown }; // output for internalFilesExports
+  protected readonly _internalFilesList?: PluginInternalFiles; // required for internalFilesExports
+
+  protected _tracer!: Tracer;
```

Comment:
or make it abstract to force the child classes to handle it

Reply:
it was designed like this previously that `moduleExports`, `tracerProvider`, `logger` and `config` will be defined during the plugin `enable`, you can't change it to be abstract as then they will have to be initialised in constructor instead of `enable`. I would break all the designs.

Reply:
hmm ok